src/backend/impl/src/services/proposal_service.rs

```diff
@@ -148,6 +179,61 @@ impl<T: ProposalRepository> ProposalServiceImpl<T> {
             proposal_repository,
         }
     }
+
+    async fn fetch_and_complete_missing_proposals(
+        &self,
+        nns_proposals: &[ProposalInfo],
+    ) -> Result<usize, ApiError> {
+        let in_progress_proposals = self
+            .proposal_repository
+            .get_proposals(Some(ReviewPeriodState::InProgress))?;
+        let missing_proposals: Vec<(ProposalId, Proposal)> = in_progress_proposals
+            .iter()
+            .filter_map(|(internal_id, proposal)| {
+                let proposal_id = proposal.nervous_system.proposal_id();
+                let is_missing = nns_proposals
+                    .iter()
+                    .all(|p| p.id.map_or(true, |nns_id| nns_id.id != proposal_id));
```

Comment:
This condition can be reversed and then used with `any` instead of `all` to create a more performant best-case runtime.

```rust
let already_exists = nns_proposals
  .iter()
  .any(|p| p.id.map_or(false, |nns_id| nns_id.id == proposal_id));
```